src/station_cmd.cpp

```diff
@@ -2506,9 +2507,16 @@ CommandCost CmdBuildDock(TileIndex tile, DoCommandFlag flags, uint32 p1, uint32
 	if (ret.Failed()) return ret;
 
 	tile_cur += TileOffsByDiagDir(direction);
-	if (!IsTileType(tile_cur, MP_WATER) || !IsTileFlat(tile_cur)) {
-		return_cmd_error(STR_ERROR_SITE_UNSUITABLE);
-	}
+
+	/* Check whether the third tile is an aqueduct head, or a water tile
+	 * with one corner raised with or without a rail track, or a buoy. */
+	if (!IsTileType(tile_cur, MP_WATER)) {
+		if (IsTileType(tile_cur, MP_TUNNELBRIDGE)) {
+			if (GetTunnelBridgeTransportType(tile_cur) != TRANSPORT_WATER) return_cmd_error(STR_ERROR_SITE_UNSUITABLE);
+		} else if (IsTileType(tile_cur, MP_RAILWAY)) {
+			if (GetRailGroundType(tile_cur != RAIL_GROUND_WATER)) return_cmd_error(STR_ERROR_SITE_UNSUITABLE);
```

Comment:
Does this work if the rail is on the side of the dock? As that sounds wrong? (I miss a directional check, basically). I could be wrong, hence the questionmark.

Reply:
The ship doesn't dock at the sides in the original code.